### PR TITLE
feat: add support for loading Docs from Steamship Files

### PR DESCRIPTION
**Pull request by douglas-reid**
Adds a Reader for persistent Steamship Files.

This provides an initial ability to use the content of Steamship Files with LlamaIndex applications. This is a stepping stone for integrating fuller support for deploying LlamaIndex in a Steamship-native manner on Steamship's platform.

### CODE REVIEW

**Comment by jerryjliu** [comment]
> Adds a Reader for persistent Steamship Files.
> 
> This provides an initial ability to use the content of Steamship Files with LlamaIndex applications. This is a stepping stone for integrating fuller support for deploying LlamaIndex in a Steamship-native manner on Steamship's platform.

thanks! will take a look soon :)

File: gpt_index/readers/steamship/file_reader.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class SteamshipFileReader(BaseReader):
+    """Reads persistent Steamship Files and converts them to Documents.
+
+    Note:
+        Requires install of `steamship` package and an active Steamship API Key.
+        To get a Steamship API Key, visit: https://steamship.com/account/api.
+        Once you have an API Key, expose it via an environment variable named
+        `STEAMSHIP_API_KEY`, or pass it as an init argument (`api_key`).
```
Comment (jerryjliu) [inline gpt_index/readers/steamship/file_reader.py:15]: nit: document api_key in Args section

File: gpt_index/readers/steamship/file_reader.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        Note:
+            The collection of Files from both `query` and `file_handles` will be
+            combined. There is no (current) support for deconflicting the collections
+            (meaning that if a file appears both in the result set of the query and
+            as a handle in file_handles, it will be loaded twice).
+        """
+        from steamship import File, Steamship
+
+        client = Steamship(workspace=workspace, api_key=self.api_key)
+        print(client)
```
Comment (jerryjliu) [inline gpt_index/readers/steamship/file_reader.py:65]: remove print

**Review by jerryjliu — COMMENTED** [review]
looks good! a few minor comments.

broader question though @douglas-reid how will this help with deploying llamaindex on steamship? this is primarily using steamship as a data loader right?

File: gpt_index/readers/steamship/file_reader.py (hunk 2 above)
Comment (douglas-reid) [reply gpt_index/readers/steamship/file_reader.py]: whoops. done.

File: gpt_index/readers/steamship/file_reader.py (hunk 1 above)
Comment (douglas-reid) [reply gpt_index/readers/steamship/file_reader.py]: done.

**Comment by douglas-reid** [comment]
> looks good! a few minor comments.
> 
> broader question though @douglas-reid how will this help with deploying llamaindex on steamship? this is primarily using steamship as a data loader right?

Thanks for the quick review.

Yes, this is only using steamship as a data loader.

I view this as a small incremental step towards running llamaindex on steamship -- though it also allows running in other places with just data from steamship too. It doesn't directly do anything special for hosting, but rather starts to enable data interchange between the two ecosystems.

It allows packages to access content that may have been created via other mechanisms to be used in a llamaindex-driven process (ex: vectorstore-powered QA). An example may be transcripts generated from lengthy uploaded video files. So, instead of rewriting any llamaindex app, you can simply use this to populate Documents, and then wrap your existing code as-is (that's the idea, anyway).

Other places where I _think_ there is value is providing access to the Steamship VectorStore and use of the Steamship LLM wrapper, depending. Those, particularly the vector store, will require a little more doing (some slight impedance mismatches in the abstractions).

Does that make sense?  Do you suggest a different approach?

**Pull request merged**